src/index.ts

```diff
@@ -1,45 +1,217 @@
+  export function isNull<T>(value: T): value is Nullable<T> {
+    return value === null;
+  }
+
+  type Nilable<T> = T extends null | undefined ? T : never;
+  export function isNil<T>(value: T): value is Nilable<T> {
+    return isNull(value) || value === undefined;
+  }
+
+  type Numberable<T> = T extends number ? T : never;
+  export function isNumber<T>(value: T): value is Numberable<T> {
+    return typeof value === 'number';
+  }
+
+  type Functionable<T> = T extends Function ? T : never;
+  export function isFunction<T>(value: T): value is Functionable<T> {
+    return typeof value === 'function';
+  }
+
+  export function shuffle<T>(value: T[]): T[] {
+    const length = value == null ? 0 : value.length;
+
+    if (!length) {
+      return [];
+    }
+
+    let index = -1;
+    const lastIndex = length - 1;
+    const result = [...value];
+
+    while (++index < length) {
+      const rand = index + Math.floor(Math.random() * (lastIndex - index + 1));
+      const value = result[rand];
+      result[rand] = result[index];
+      result[index] = value;
+    }
+
+    return result;
+  }
+
+  export function pick<T extends Record<string, unknown>, R extends keyof T>(
+    obj: T,
+    ...keys: R[]
+  ): Pick<T, R> {
+    return keys.reduce((acc, key) => {
+      acc[key] = obj[key];
+
+      return acc;
+    }, {} as Pick<T, R>);
+  }
+
+  export function omit<T extends Record<string, unknown>, R extends keyof T>(
+    obj: T,
+    ...keys: R[]
+  ): Omit<T, R> {
+    return keys.reduce(
+      (acc, key) => {
+        delete acc[key];
 
-	export function isNull() {}
+        return acc;
+      },
+      { ...obj }
+    );
+  }
 
-	export function isNil() {}
+  export function memoize<T extends (...args: any[]) => any>(
+    func: T,
+    makeKey: (...args: any[]) => string
+  ): (...args: any[]) => any {
+    const cache: Record<string, any> = {};
 
-	export function isNumber() {}
+    return (...args: any[]): any => {
+      const key = makeKey(args);
 
-	export function isFunction() {}
+      return cache[key] ? cache[key] : (cache[key] = func(...args));
+    };
+  }
 
-	export function shuffle() {}
+  export function debounce<T extends (...args: any[]) => void>(
+    func: T,
+    wait: number
+  ): (...args: any[]) => void {
+    let timer: ReturnType<typeof setTimeout> | null = null;
 
-	export function pick() {}
+    return (...args: any[]) => {
+      if (timer) clearTimeout(timer);
 
-	export function omit() {}
+      timer = setTimeout(() => func(...args), wait);
+    };
+  }
 
-	export function memoize() {}
+  export function throttle<T extends (...args: any[]) => void>(
+    func: T,
+    wait: number
+  ): (...args: any[]) => void {
+    let timer: ReturnType<typeof setTimeout> | null = null;
 
-	export function debounce() {}
+    return (...args: any[]) => {
+      if (!timer) {
+        timer = setTimeout(() => {
+          timer = null;
+          func(...args);
+        }, wait);
+      }
+    };
+  }
```

Comment:
리턴되는 함수의 args 가 결국 `func` 매개변수의 args 로 들어가네요!
까다롭겠지만 두 타입을 어떻게든 일치시켜보는거 어떠신가요 !!
저라면 아래와 같이 할 것 같아요 

```suggestion
   export function throttle<T extends (...args: any[]) => void>(
    func: T,
    wait: number
  ): (...args: Parameters<T>) => void {
    let timer: ReturnType<typeof setTimeout> | null = null;

    return (...args: Parameters<T>) => {
      if (!timer) {
        timer = setTimeout(() => {
          timer = null;
          func(...args);
        }, wait);
      }
    };
  }

```

Reply:
이것도 어렵네요 ㅋㅋㅋㅋㅋㅋㅋㅋㅋ 호프 옆에서 물어보면서 고칠게요!